### PR TITLE
feat(noise): add Noise Handshake processing and after-handshake encryption

### PR DESCRIPTION
**Pull request by s1fr0**
This PR introduces data structures and procedures required to execute Noise Handshakes between Waku users according to [Noise specification](http://www.noiseprotocol.org/noise.html#processing-rules). 

Specifically, it adds:
- Generic processing of handshake pre-message and message patterns (`processPreMessagePatternTokens`, `processMessagePatternPayload`, `processMessagePatternTokens`). To add support to a new handshake, it suffices to add its definition in `NoiseHandshakePatterns` and add for it a new protocol ID in `PayloadV2ProtocolIDs`;
- Generic 1 step handshake advancement (`stepHandshake`): given current user's state, processes the next handshake message.
- Handshake finalization and symmetric state split (`finalizeHandshake`);
- Procedures for after-handshake inbound/outbound symmetric encryption (`writeMessage`, `readMessage`);
- Some tests for each supported handshake.

This PR extends https://github.com/status-im/nim-waku/pull/933 (which in turn extends https://github.com/status-im/nim-waku/pull/932, https://github.com/status-im/nim-waku/pull/931 and https://github.com/status-im/nim-waku/pull/930) and completes all the non-optional points of https://github.com/status-im/nim-waku/issues/881.

### CODE REVIEW

**Comment by s1fr0** [comment]
@staheri14 @jm-clius @kaiserd ready for review! Thanks!

File: waku/v2/protocol/waku_noise/noise.nim [hunk 1]
```diff
@@ -917,4 +923,565 @@ proc deserializePayloadV2*(payload: seq[byte]): Result[PayloadV2, cstring]
...
\ No newline at end of file
+  return ok(payload2)
+
+
+#################################################################
+
+# Handshake Processing
+
+#################################
+## Utilities
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:934]: @jm-clius @staheri14 After internal discussion, we decided to not proceed (at the moment) with the split of the module in smaller submodules. Reasons for this include: 
- the necessity to make the majority of object private fields public; 
- most of already merged code will appear in the PR diff, thus making identifying and reviewing new parts difficult.

After all Noise features are implemented and merged, we can proceed with splitting the module in smaller submodules. By that time we might support already [import foo {.all.}](https://github.com/nim-lang/Nim/pull/17706) directive, that will ease such refactoring.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 2]
```diff
@@ -917,4 +923,565 @@ proc deserializePayloadV2*(payload: seq[byte]): Result[PayloadV2, cstring]
...
+  hs.s = staticKey
+  hs.psk = psk
+  hs.msgPatternIdx = 0
+  hs.initiator = initiator
+  # We process any eventual handshake pre-message pattern by processing pre-message public keys
+  processPreMessagePatternTokens(hs, preMessagePKs)
+  return hs
+
+# Advances 1 step in handshake
+proc stepHandshake*(rng: var BrHmacDrbgContext, hs: var HandshakeState, readPayloadV2: PayloadV2 = default(PayloadV2), transportMessage: seq[byte] = @[]): Result[HandshakeStepResult, cstring]
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:1365]: As per our call, please add a comment about when `readPaylodV2` and `transportMessage`  should be passed.

File: tests/v2/test_waku_noise.nim [hunk 3]
```diff
@@ -413,3 +412,438 @@ procSuite "Waku Noise":
...
+      bobStep.transportMessage == sentTransportMessage
+
+    # Note that for this handshake pattern, no more message patterns are left for processing 
+    # Another call to stepHandshake would return an empty HandshakeStepResult
+    # We test that extra calls to stepHandshake do not affect handshake finalization
+    bobStep = stepHandshake(rng[], bobHS, transportMessage = sentTransportMessage).get()
+    aliceStep = stepHandshake(rng[], aliceHS, readPayloadV2 = bobStep.payload2).get()
+    aliceStep = stepHandshake(rng[], aliceHS, transportMessage = sentTransportMessage).get()
+    bobStep = stepHandshake(rng[], bobHS, readPayloadV2 = aliceStep.payload2).get()
+
```
Comment (staheri14) [inline tests/v2/test_waku_noise.nim:490]: As per our call, please add an explicit check to  whether the handshake states are remained unchanged or not. The same for the hand shake results.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 4]
```diff
@@ -917,4 +923,565 @@ proc deserializePayloadV2*(payload: seq[byte]): Result[PayloadV2, cstring]
...
+      hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+  return ok(outHandshakeMessage)
+
+#################################
+## Procedures to progress handshakes between users
+#################################
+
+# Initializes a Handshake State
+proc initialize*(hsPattern: HandshakePattern, staticKey: KeyPair = default(KeyPair), prologue: seq[byte] = @[], psk: seq[byte] = @[], preMessagePKs: seq[NoisePublicKey] = @[], initiator: bool = false): HandshakeState
```
Comment (staheri14) [inline waku/v2/protocol/waku_noise/noise.nim:1352]: As per our 1:1, you may add an ephemeral keypair as a parameter here to account for pre-shared ephemeral keys

**Review by staheri14 — COMMENTED** [review]
Good job! I have added some comments mostly covering what we have discussed in our 1:1 (just as a placeholder).

**Comment by staheri14** [comment]
Also, the macOS CI checks are failing. @s1fr0

File: tests/v2/test_waku_noise.nim [hunk 5]
```diff
@@ -321,7 +320,7 @@ procSuite "Waku Noise":
     ########################################
 
     # We generate random input key material and we execute mixKey
-    var inputKeyMaterial = randomChaChaPolyKey(rng[])
+    var inputKeyMaterial = randomSeqByte(rng[], rand(1..128))
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim:323]: No, this was a bug coming from [libp2p noise](https://github.com/status-im/nim-libp2p/blob/cba3ca3c3e7bb0ffa8d91aa7c3115a90eb17712a/libp2p/protocols/secure/noise.nim#L183) implementation, where it is still present. Input key material can be arbitrary input and not a only a ChaChaPolyKey.

File: waku/v2/protocol/waku_noise/noise.nim [hunk 6]
```diff
@@ -152,13 +152,21 @@ type
     msgPatternIdx: uint8
     psk: seq[byte]
 
+  # While processing messages patterns, users either:
+  # - read (decrypt) a the other party's (encrypted) transport message
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/934/commits/1d86c99113c3a79e07892e5cf8821baf104845a3

File: tests/v2/test_waku_noise.nim [hunk 7]
```diff
@@ -413,3 +412,438 @@ procSuite "Waku Noise":
...
       getNonce(cs2) == 0.uint64
       getKey(cs1) != getKey(cs2)
+
+  test "Noise XX Handhshake and message encryption (extended test)":
+
+    let hsPattern = NoiseHandshakePatterns["XX"]
+
+    # We initialize Alice's and Bob's Handshake State
+    let aliceStaticKey = genKeyPair(rng[])
+    var aliceHS = initialize(hsPattern, staticKey = aliceStaticKey, initiator = true)
```
[the remark this answers is not in the document]
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim]: Addressed in https://github.com/status-im/nwaku/pull/934/commits/1d86c99113c3a79e07892e5cf8821baf104845a3

File: waku/v2/protocol/waku_noise/noise.nim (hunk 2 above)
Comment (s1fr0) [reply waku/v2/protocol/waku_noise/noise.nim:1369]: Addressed in https://github.com/status-im/nwaku/pull/934/commits/1d86c99113c3a79e07892e5cf8821baf104845a3

File: tests/v2/test_waku_noise.nim (hunk 3 above)
Comment (s1fr0) [reply tests/v2/test_waku_noise.nim:494]: Addressed in https://github.com/status-im/nwaku/pull/934/commits/1d86c99113c3a79e07892e5cf8821baf104845a3

**Review by staheri14 — APPROVED** [review]
LGTM!

**Pull request merged**